### PR TITLE
Add PSO caching for advanced indexing kernels

### PR DESCRIPTION
**Pull request by DenisVieriu97**
Use bindless Argument Buffers / unbounded arrays from Metal 3 to write directly in the AB for advanced indexing kernels. This allows the PSO to be cached since we can fill directly the AB on the CPU.

### CODE REVIEW

**Comment by ssaladis** [comment]
Why are the PSOs not being cached ? It shouldn't be related to not using ABs. Is it because binding point is changing so PSO is not cached ? So move to ABs which are bindless ?

**Comment by DenisVieriu97** [comment]
> Why are the PSOs not being cached ? It shouldn't be related to not using ABs. Is it because binding point is changing so PSO is not cached ? So move to ABs which are bindless ?

This PR is using bindless ABs directly from Metal 3 (the PSO will be cached). 

>  Is it because binding point is changing so PSO is not cached ? 
Yes, that was the cause previous the bindless ABs

**Review by razarmehr — APPROVED** [review]
LGTM

File: aten/src/ATen/mps/IndexKernels.h [hunk 1]
```diff
@@ -9,27 +9,42 @@ static const char * indexing_metal_shaders = R"INDEX_METAL(
 
 using namespace metal;
 
-constant uint32_t num_indices            [[function_constant(0)]];
-
+#if __METAL_VERSION__ < 300
```
Comment (kulinseth) [inline aten/src/ATen/mps/IndexKernels.h:12]: Lets be consistent and move this to `__METAL_VERSION__ >= 300`

**Review by kulinseth — COMMENTED** [review]
This code needs to be simplified, do we need the AB logic , did it help with performance on torchbench?

File: aten/src/ATen/mps/IndexKernels.h (hunk 1 above)
Comment (kulinseth) [reply aten/src/ATen/mps/IndexKernels.h:12]: Please clean it up as followup.

**Pull request merged**